src/poetry/puzzle/provider.py

```diff
@@ -546,35 +546,53 @@ def complete_package(
             package = dependency_package.package
             dependency = dependency_package.dependency
             requires = package.all_requires
-        elif package.source_type not in {
-            "directory",
-            "file",
-            "url",
-            "git",
-        }:
-            try:
-                dependency_package = DependencyPackage(
-                    dependency,
-                    self._pool.package(
-                        package.pretty_name,
-                        package.version,
-                        extras=list(dependency.extras),
-                        repository=dependency.source_name,
-                    ),
-                )
-            except PackageNotFound as e:
+        else:
+            if package.source_type == "file":
+                if len(package.files) > 1:
+                    # a locked package might contain further files
+                    # for multiple-constraints dependencies -> prune
+                    path = urllib.parse.urlsplit(package.source_url).path
+                    assert isinstance(path, str)
+                    filename = Path(path).name
+                    for item in package.files:
+                        if item["file"] == filename:
+                            package.files = [item]
+                            break
+                    else:
+                        self.debug(
+                            f"<debug>No entry for {package} in metadata files of lock"
+                            " file</debug>"
+                        )
+                        package = self.search_for_direct_origin_dependency(
+                            package.to_dependency()
+                        )
+            elif package.source_type in {"directory", "git", "url"}:
+                # these types don't have files, however,
+                # a locked package might contain further files
+                # for multiple-constraints dependencies -> clear
+                package.files = []
+            else:
                 try:
-                    dependency_package = next(
-                        DependencyPackage(dependency, pkg)
-                        for pkg in self.search_for_installed_packages(dependency)
+                    dependency_package = DependencyPackage(
+                        dependency,
+                        self._pool.package(
+                            package.pretty_name,
+                            package.version,
+                            extras=list(dependency.extras),
+                            repository=dependency.source_name,
+                        ),
                     )
-                except StopIteration:
-                    raise e from e
+                except PackageNotFound as e:
+                    try:
+                        dependency_package = next(
```

Comment:
I know this is a generator expression and thus evaluated lazily, but it seems more straightforward to construct this such that the `next()` is called on `self.search_for_installed_packages` and we make the `DependencyPackage` from the result of that call.

Also, is there a reason `pkg.to_dependency()` wouldn't work here?

Reply:
I haven't changed anything here, just restructured the `if-else`. I don't want to include unrelated cleanups here because as discussion shows this fix will not survive long on master.